concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+                    }
+
+                    Ok(v1::ReceiveResult::Success {
+                        logs,
+                        state_changed,
+                        return_value,
+                        remaining_energy,
+                    })
+                }
+                v1::ReceiveResult::Interrupt {
+                    remaining_energy,
+                    state_changed,
+                    logs,
+                    config,
+                    interrupt,
+                } => {
+                    // Create the interrupt event, which will be included for transfers, calls, and
+                    // upgrades, but not for the remaining interrupts.
+                    let interrupt_event = ChainEvent::Interrupted {
+                        address: self.address,
+                        logs,
+                    };
+                    match interrupt {
+                        v1::Interrupt::Transfer { to, amount } => {
+                            // Add the interrupt event
+                            self.chain_events.push(interrupt_event);
+
+                            let response = match self
+                                .invocation_handler
+                                .transfer_from_contract_to_account(amount, self.address, to)
+                            {
+                                Ok(new_balance) => v1::InvokeResponse::Success {
+                                    new_balance,
+                                    data: None,
+                                },
+                                Err(err) => {
+                                    let kind = match err {
+                                        TransferError::ToMissing => {
+                                            v1::InvokeFailure::NonExistentAccount
+                                        }
+                                        TransferError::InsufficientBalance => {
+                                            v1::InvokeFailure::InsufficientAmount
+                                        }
+                                    };
+                                    v1::InvokeResponse::Failure { kind }
+                                }
+                            };
+
+                            let success = matches!(response, v1::InvokeResponse::Success { .. });
+                            if success {
+                                // Add transfer event
+                                self.chain_events.push(ChainEvent::Transferred {
+                                    from: self.address,
+                                    amount,
+                                    to,
+                                });
+                            }
+                            // Add resume event
+                            self.chain_events.push(ChainEvent::Resumed {
+                                address: self.address,
+                                success,
+                            });
```

Comment:
Maybe I am misunderstanding something, but I can't see where these events are rolled back if the entire contract invocation fails. Can you point that out?

Reply:
They are only appended on success. It happens in the bottom of `invoke_entrypoint`
```
        // Append the new chain events if the invocation succeeded.
        if result.is_success() {
            chain_events.append(&mut new_chain_events);
        }
```